python/example_code/rds/lending_library/library_api/chalicelib/library_data.py

```diff
@@ -0,0 +1,397 @@
+            logger.exception("Run batch statement on %s failed.", self._db_name)
+            raise
+        else:
+            return result
+
+    def bootstrap_tables(self):
+        """
+        Creates tables in the database. The tables are defined in the constructor.
+        """
+        for table in self._tables.values():
+            logger.info("Creating table %s.", table.name)
+            sql = create_table(table)
+            self._run_statement(sql)
+
+    def add_books(self, books):
+        """
+        Adds a list of books and their authors to the database. The list of authors
+        is first processed to remove duplicates.
+
+        :param books: The list of books and their authors to add to the database.
+        :return: The counts of authors and books added to the database.
+        """
+        authors = {
+            book['author']: {
+                'FirstName': ' '.join(book['author'].split(' ')[:-1]),
+                'LastName': book['author'].split(' ')[-1]
+            } for book in books
+        }
+        sql, sql_param_sets = insert(
+            self._tables['Authors'], authors.values())
+        result = self._run_batch_statement(sql, sql_param_sets)
+        author_count = len(result['updateResults'])
+        logger.info("Added %s authors to the database.", author_count)
+
+        auth_ids = [field['generatedFields'][0]['longValue']
+                    for field in result['updateResults']]
+        for auth, auth_id in zip(authors.values(), auth_ids):
+            auth['author_id'] = auth_id
+        sql, sql_param_sets = insert(
+            self._tables['Books'], [{
+                'Title': book['title'],
+                'AuthorID': authors[book['author']]['author_id']
+            } for book in books])
+        result = self._run_batch_statement(sql, sql_param_sets)
+        book_count = len(result['updateResults'])
+        logger.info("Added %s books to the database.", book_count)
+        return author_count, book_count
+
+    def get_books(self, author_id=None):
+        """
+        Gets books from the database.
+
+        :param author_id: When specified, only books by this author are returned.
+                          Otherwise, all books are returned.
+        :returns: The list of books.
+        """
+        logger.info("Listing by author %s.", "All" if author_id is None else author_id)
+        where_clauses = None if author_id is None else [{
+            'table': 'Authors', 'column': 'AuthorID', 'op': '=',
+            'value': author_id}]
+        sql, columns, params = query('Books', self._tables, where_clauses)
+        results = self._run_statement(sql, sql_params=params)
+        output = unpack_query_results(columns, results)
+        return output
+
+    def add_book(self, book):
+        """
+        Adds a book and its author to the database. This functions uses a database
+        transaction to ensure that both the author and the book are added if one
```

Comment:
read this: "transaction to ensure that both the author and the book are added if one
        of the inserts fails, the transaction is rolled back and nothing is
        added." Should it be "that if both the author and the book are added, and if one of the inserts fails, the transaction..."?